gilt/git.py

```diff
@@ -29,6 +30,37 @@
 from gilt import util
 
 
+def _git_env(repository):
+    """
+    Git environment variables for specific repository.
+
+    :param repository: path to a git repository.
+    :return: dict of envionment variables.
+    """
+    return {
+        "GIT_WORK_TREE": repository,
+        "GIT_DIR": os.path.join(repository, ".git"),
+        "GIT_OBJECT_DIRECTORY": os.path.join(repository, ".git", "objects")
+    }
+
+
+def do_overlay(config, debug=False):
+    """
+    Actual methods the overlays git based repo.
+
+    :param config: A dict contains configuration of the git resource.
+    :param debug: An optional bool to toggle debug output.
+    :return: None
+    """
+    with fasteners.InterProcessLock(config.lock_file):
```

Comment:
Using this with threads is a problem.  The InterProcessLock is not meant to be used by threads from the same process.  See http://fasteners.readthedocs.io/en/latest/api/process_lock.html.  You can check that it doesn't work by running: https://gist.github.com/rabadin/f39e5b097583b65ee1ef9f952d0ff948.

This creates conflicts when you run this on a gilt.yml file that uses the *same* source.  If you run it on ans-eng's gilt.yml it will break with:
```
  STDERR:
fatal: Unable to create '/Users/rabadin/.gilt/clone/metacloud.ansible-systems/.git/index.lock': File exists.

Another git process seems to be running in this repository, e.g.
an editor opened by 'git commit'. Please make sure all processes
are terminated then try again. If it still fails, a git process
may have crashed in this repository earlier:
remove the file manually to continue.
```